operator_manifest/resolver.py

```diff
@@ -0,0 +1,131 @@
+# SPDX-License-Identifier: BSD-3-Clause
+import functools
+import hashlib
+import json
+import logging
+import os.path
+import subprocess
+
+
+logger = logging.getLogger(__name__)
+
+
+def resolve_image_reference(image_reference, authfile=None):
+    """
+    Resolve the image reference to a digest image reference.
+
+    :param str image_reference: the image reference of the container image to resolve
+    :param str authfile: the path of the authentication file for registry communication.
+    :return: the image reference resolved to a digest
+    :rtype: str
+    """
+    extra_args = []
+    if authfile:
+        if not os.path.exists(authfile):
+            raise ValueError(f'Specified authfile {authfile} does not exist')
+        extra_args.append('--authfile')
+        extra_args.append(authfile)
+
+    logger.debug('Resolving %s', image_reference)
+    name = _get_container_image_name(image_reference)
+    skopeo_raw = _skopeo_inspect(f'docker://{image_reference}', '--raw', *extra_args)
+    if json.loads(skopeo_raw).get('schemaVersion') == 2:
+        raw_digest = hashlib.sha256(skopeo_raw.encode('utf-8')).hexdigest()
+        digest = f'sha256:{raw_digest}'
+    else:
+        # Schema 1 is not a stable format. The contents of the manifest may change slightly
+        # between requests causing a different digest to be computed. Instead, let's leverage
+        # skopeo's own logic for determining the digest in this case. In the future, we
+        # may want to use skopeo in all cases, but this will have significant performance
+        # issues until https://github.com/containers/skopeo/issues/785
+        digest = json.loads(_skopeo_inspect(f'docker://{image_reference}', *extra_args))['Digest']
+    resolved_image_reference = f'{name}@{digest}'
+    logger.debug('%s resolved to %s', image_reference, resolve_image_reference)
+    return resolved_image_reference
+
+
+def _get_container_image_name(image_reference):
+    """
+    Get the container image name from an image reference.
+
+    :param str image_reference: the image reference to analyze
+    :return: the container image name
+    """
+    if '@' in image_reference:
+        return image_reference.split('@', 1)[0]
+    else:
+        return image_reference.rsplit(':', 1)[0]
+
+
+def _retry(attempts=3, wait_on=Exception):
+    """
+    Retry a section of code until success or max attempts are reached.
```

Comment:
"section of code" may be misleading, maybe "function"? This made me think if, through some dark magic, one could create a retry *context manager* :thinking:

Reply:
Changed it to `function`. Also clarified in the docstring that this is actually a decorator.

Making a retry context manager should be pretty straight forward, I think. Although, for this case, a decorator is probably just fine?

Reply:
Yeah, a decorator is definitely the way to go, I was just thinking out loud. I'm not actually sure if it's possible, since
a) a context manager does not know what code is being executed inside the context (unless you can somehow find that in the traceback that is passed to `__exit__`)
b) a context ends immediately with the first exception